tls/s2n_connection.c

```diff
@@ -1392,3 +1397,49 @@ int s2n_connection_set_keyshare_by_name_for_testing(struct s2n_connection *conn,
 
     POSIX_BAIL(S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
 }
+
+int s2n_connection_get_peer_cert_chain(const struct s2n_connection *conn, struct s2n_cert_chain_and_key *cert_chain_and_key)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(cert_chain_and_key);
+
+    const struct s2n_x509_validator *validator = &conn->x509_validator;
+    POSIX_ENSURE_REF(validator);
+    POSIX_ENSURE(validator->state == VALIDATED, S2N_ERR_CERT_NOT_VALIDATED);
+
+    struct s2n_cert_chain *cert_chain = cert_chain_and_key->cert_chain;
+    POSIX_ENSURE_REF(cert_chain);
+    struct s2n_cert **insert = &cert_chain->head;
+    POSIX_ENSURE(*insert == NULL, S2N_ERR_INVALID_CERT_CHAIN);
+
+    /* X509_STORE_CTX_get1_chain() returns a validated cert chain if a previous call to X509_verify_cert() was successful.
+     * X509_STORE_CTX_get0_chain() is a better API because it doesn't return a copy. But it's not available for Openssl 1.0.2.
+     * See the comments here:
+     * https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get1_chain.html
+     */
+    STACK_OF(X509) *cert_chain_validated = X509_STORE_CTX_get1_chain(validator->store_ctx);
+    POSIX_ENSURE_REF(cert_chain_validated);
+
+    for (size_t cert_idx = 0; cert_idx < sk_X509_num(cert_chain_validated); cert_idx++) {
+        X509 *cert = sk_X509_value(cert_chain_validated, cert_idx);
+        uint8_t *cert_data = NULL;
```

Comment:
`cert_data` will leak if any of the GUARDs our ENSUREs return early. Use `DEFER_CLEANUP` instead

Reply:
Yes, this is what I was trying to explain in the call-out section. 

Here, the memory allocation is done within the `i2d_X509` API call. 
```c
int cert_size = i2d_X509(cert, &cert_data);
```

and we cannot call DEFER_CLEANUP for the `OPENSSL_free(cert_data);` free here as DEFER_CLEANUP is defined as:
```c
#define DEFER_CLEANUP(_thealloc, _thecleanup) \
   __attribute__((cleanup(_thecleanup))) _thealloc
```

We can't do: 

```c
DEFER_CLEANUP(int cert_size = i2d_X509(cert, &cert_data), OPENSSL_free);
```

I could use a temporary blob maybe and free that blob? But that also requires me to free using OPENSSL_free and not s2n_free.

Reply:
I'm not sure why you can't use defer cleanup. Wouldn't it just be:

```c
DEFER_CLEANUP(uint8_t *cert_data = NULL, OPENSSL_free);
int cert_size = i2d_X509(cert, &cert_data);
POSIX_ENSURE_GT(cert_size, 0);
```

Reply:
The first argument to `DEFER_CLEANUP` is expected to be a function or static declaration. 

I get the following error message for the above code: 

```c
s2n_connection.c:1425:9: error: cleanup argument not a function
         DEFER_CLEANUP(uint8_t *cert_data = NULL, OPENSSL_free);
```

Reply:
I get the following error message for the above code: 

```c
s2n_connection.c:1425:9: error: cleanup argument not a function
         DEFER_CLEANUP(uint8_t *cert_data = NULL, OPENSSL_free);
```

Reply:
Turns out `OPENSSL_free` is a macro... We'll need to use `DEFINE_POINTER_CLEANUP_FUNC`.

Reply:
Fun! 
```c
# define OPENSSL_free(addr) \
        CRYPTO_free(addr, OPENSSL_FILE, OPENSSL_LINE)
 ```